### PR TITLE
WT-10241 Creating a CppSuite workload that stresses the reverse split code path

### PR DESCRIPTION
**Pull request by luke-pearson**
I've added a new `barrier` class here intended to sychronise thread workers as they gradually drift from eachother when having the same op_rate.

### CODE REVIEW

File: test/cppsuite/src/util/barrier.h [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+/*
+ * A barrier is a sychronization class that allows i_count threads to perform an action at the same
+ * time. They are introduced in C++20 so we need a basic implementation in the meantime.
+ *
+ * In order to synchronize across threads call wait() on each thread, once i_count threads have
+ * called wait() they will exit the wait() and continue.
+ */
+class barrier {
+    public:
+    explicit barrier(std::size_t i_count);
```
Comment (jeremythorp) [inline test/cppsuite/src/util/barrier.h:45]: I'd recommend adding 

`barrier(barrier const&) = default;`
`~barrier() = default;`

to clarify that the default copy constructor and destructor are correct.

File: test/cppsuite/src/util/barrier.h [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+ * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#ifndef BARRIER_H
```
Comment (jeremythorp) [inline test/cppsuite/src/util/barrier.h:29]: The `#ifndef` include guards can now be done with `#pramga once`

File: test/cppsuite/tests/reverse_split.cpp [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+                continue;
+            }
+            const char *key;
+            testutil_check(write_cursor->get_key(write_cursor.get(), &key));
+            const std::string key_str(key);
+            uint64_t min_key_id = std::stoi(key_str);
+            uint64_t key_count = coll.get_key_count();
+            /* Truncate up to 80% of the range. */
+            uint64_t end_key_id = random_generator::instance().generate_integer<uint64_t>(
+              min_key_id, min_key_id + ((key_count - min_key_id) / 1.2));
```
Comment (jeremythorp) [inline test/cppsuite/tests/reverse_split.cpp:84]: Dividing by 1.2 will give slightly more that 80% (about 83%).

**Review by jeremythorp — APPROVED** [review]
I have made a few suggestions, but otherwise LGTM!

File: test/cppsuite/configs/reverse_split_stress.txt [hunk 4]
```diff
@@ -0,0 +1,46 @@
+# This workload is inserts content continuously and truncates off the start of the file.
+# The truncations truncate up to 80% of the file each time.
```
Comment (etienneptl) [inline test/cppsuite/configs/reverse_split_stress.txt:2]: Should we explain how much the stress configuration differs from the default one? It's hard to go back and forth to compare the two and find the differences.

File: test/cppsuite/src/main/test.cpp [hunk 5]
```diff
@@ -128,7 +128,7 @@ test::run()
     db_create_config += ",cache_max_wait_ms=" + std::to_string(cache_max_wait_ms);
 
     /* Add the user supplied wiredtiger open config. */
-    db_create_config += _args.wt_open_config;
+    db_create_config += "," + _args.wt_open_config;
```
Comment (etienneptl) [inline test/cppsuite/src/main/test.cpp:131]: Nice one!

File: test/cppsuite/tests/reverse_split.cpp [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+ * This test inserts data at the end of the collection and truncates off from the start of the
+ * collection. In doing so pages at the start of the tree are gradually emptied while pages are
+ * added at the end of the tree. This means the test frequently executes the reverse split path.
+ */
+class reverse_split : public test {
+    public:
+    reverse_split(test_args &args) : test(args)
+    {
+        /* Add split timing stresses to the conn_open config. */
+        if (args.wt_open_config == "") {
```
Comment (etienneptl) [inline test/cppsuite/tests/reverse_split.cpp:44]: Change to `if (args.wt_open_config.empty()) {`

File: test/cppsuite/tests/reverse_split.cpp [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+/*
+ * This test inserts data at the end of the collection and truncates off from the start of the
+ * collection. In doing so pages at the start of the tree are gradually emptied while pages are
+ * added at the end of the tree. This means the test frequently executes the reverse split path.
+ */
+class reverse_split : public test {
+    public:
+    reverse_split(test_args &args) : test(args)
+    {
+        /* Add split timing stresses to the conn_open config. */
```
Comment (etienneptl) [inline test/cppsuite/tests/reverse_split.cpp:43]: Why do we do this only if `wt_open_config` is empty?

File: test/cppsuite/tests/reverse_split.cpp [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+                /* A return value of true implies the insert was successful. */
+                if (!tc->insert(write_cursor, coll.id, key, value)) {
+                    tc->txn.rollback();
+                    break;
+                }
+                added_count++;
+                if (tc->txn.can_commit()) {
+                    if (tc->txn.commit()) {
+                        coll.increase_key_count(added_count);
+                        start_key = coll.get_key_count();
```
Comment (etienneptl) [inline test/cppsuite/tests/reverse_split.cpp]: Do you need this? As soon as you commit, `tc->txn.active()` is false and the `while` loop will exit.

File: test/cppsuite/tests/reverse_split.cpp [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+            tc->txn.begin();
+            int ret = write_cursor->next(write_cursor.get());
+            if (ret != 0) {
+                tc->txn.rollback();
+                continue;
+            }
+            const char *key;
+            testutil_check(write_cursor->get_key(write_cursor.get(), &key));
+            const std::string key_str(key);
+            /* Truncate up to 20% of the range. */
```
Comment (etienneptl) [inline test/cppsuite/tests/reverse_split.cpp]: You mention 80% at other places

File: test/cppsuite/tests/run.cpp [hunk 10]
```diff
@@ -46,6 +46,7 @@
 #include "search_near_02.cpp"
 #include "search_near_03.cpp"
 #include "test_template.cpp"
+#include "reverse_split.cpp"
```
Comment (etienneptl) [inline test/cppsuite/tests/run.cpp:49]: I know it's annoying and it's not from you but.. A->Z please

File: test/cppsuite/tests/run.cpp [hunk 11]
```diff
@@ -163,6 +164,8 @@ run_test(const std::string &test_name, const std::string &config, const std::str
         search_near_03(args).run();
     else if (test_name == "test_template")
         test_template(args).run();
+    else if (test_name == "reverse_split")
```
Comment (etienneptl) [inline test/cppsuite/tests/run.cpp:167]: A->Z

File: test/cppsuite/tests/run.cpp [hunk 12]
```diff
@@ -186,7 +189,7 @@ main(int argc, char *argv[])
     std::string cfg, config_filename, current_cfg, current_test_name, home, test_name,
       wt_open_config;
     int64_t error_code = 0;
-    const std::vector<std::string> all_tests = {"bounded_cursor_perf",
+    const std::vector<std::string> all_tests = {"reverse_split", "bounded_cursor_perf",
```
Comment (etienneptl) [inline test/cppsuite/tests/run.cpp:192]: A->Z

File: test/cppsuite/src/util/barrier.cpp [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+
+namespace test_harness {
+barrier::barrier(std::size_t i_count) : _threshold(i_count), _count(i_count), _generation(0) {}
+
+void
+barrier::wait()
+{
+    std::unique_lock<std::mutex> lock{_mutex};
+    auto lock_gen = _generation;
+    if (!--_count) {
```
Comment (etienneptl) [inline test/cppsuite/src/util/barrier.cpp:39]: isnt it dangerous not the check if `_count` is `0`?

File: test/cppsuite/src/component/workload_manager.cpp [hunk 14]
```diff
@@ -88,10 +90,13 @@ workload_manager::run()
             logger::log_msg(LOG_INFO,
               "workload_manager: Creating " + std::to_string(it.thread_count) + " " +
                 type_string(it.type) + " threads.");
+        /* Create a synchronisation object. */
+        std::shared_ptr<barrier> barrier_ptr = std::make_shared<barrier>(it.thread_count);
```
Comment (etienneptl) [inline test/cppsuite/src/component/workload_manager.cpp:94]: What issues did you face to need this?

File: test/cppsuite/tests/reverse_split.cpp [hunk 15]
```diff
@@ -128,6 +65,8 @@ class reverse_split : public test {
         scoped_cursor write_cursor = std::move(tc->session.open_scoped_cursor(coll.name));
 
         while (tc->running()) {
+            /* Synchronise across our truncate threads so they don't diverge over time. */
```
Comment (etienneptl) [inline test/cppsuite/tests/reverse_split.cpp:68]: What do you mean by diverge?

Comment (luke-pearson) [reply test/cppsuite/tests/reverse_split.cpp:68]: Lets say there are 5 truncate threads running, they will all sleep the specified amount every iteration however their actions can take different amounts of time. So when the test begins they all call truncate at the same time and by the time the test ends they are calling truncate at varied times. 

The synchronize code ensure they all call truncate at roughly the same time for the duration of the test which results in a spikier workload.

File: test/cppsuite/tests/reverse_split.cpp (hunk 7 above)
Comment (luke-pearson) [reply test/cppsuite/tests/reverse_split.cpp:43]: I will update the comment to explain.

File: test/cppsuite/tests/reverse_split.cpp (hunk 8 above)
Comment (luke-pearson) [reply test/cppsuite/tests/reverse_split.cpp]: I think you reviewed an older version, this code is removed in later commits.

File: test/cppsuite/tests/reverse_split.cpp (hunk 9 above)
Comment (luke-pearson) [reply test/cppsuite/tests/reverse_split.cpp]: Wrong version of the code.

File: test/cppsuite/src/component/workload_manager.cpp (hunk 14 above)
Comment (luke-pearson) [reply test/cppsuite/src/component/workload_manager.cpp:94]: See my diverge answer below.

File: test/cppsuite/src/util/barrier.h (hunk 1 above)
Comment (luke-pearson) [reply test/cppsuite/src/util/barrier.h:47]: 87c1d8abf

File: test/cppsuite/src/util/barrier.h (hunk 2 above)
Comment (luke-pearson) [reply test/cppsuite/src/util/barrier.h]: 87c1d8abf

File: test/cppsuite/tests/reverse_split.cpp (hunk 3 above)
Comment (luke-pearson) [reply test/cppsuite/tests/reverse_split.cpp:92]: 87c1d8abf

File: test/cppsuite/configs/reverse_split_stress.txt (hunk 4 above)
Comment (luke-pearson) [reply test/cppsuite/configs/reverse_split_stress.txt]: 87c1d8abf

File: test/cppsuite/tests/reverse_split.cpp (hunk 6 above)
Comment (luke-pearson) [reply test/cppsuite/tests/reverse_split.cpp]: 87c1d8abf

File: test/cppsuite/tests/reverse_split.cpp (hunk 7 above)
Comment (luke-pearson) [reply test/cppsuite/tests/reverse_split.cpp]: 87c1d8abf

File: test/cppsuite/src/util/barrier.cpp (hunk 13 above)
Comment (luke-pearson) [reply test/cppsuite/src/util/barrier.cpp:39]: You mean prior to the decrement? We are guaranteed for _count to be reset to _threshold if it reaches 0.

File: test/cppsuite/tests/run.cpp (hunk 12 above)
Comment (luke-pearson) [reply test/cppsuite/tests/run.cpp:192]: [6302c61](https://github.com/wiredtiger/wiredtiger/pull/8554/commits/6302c6171dea664d9b428e946d44ef15a948d742)

File: test/cppsuite/tests/run.cpp (hunk 11 above)
Comment (luke-pearson) [reply test/cppsuite/tests/run.cpp]: [6302c61](https://github.com/wiredtiger/wiredtiger/pull/8554/commits/6302c6171dea664d9b428e946d44ef15a948d742)

**Review by etienneptl — APPROVED** [review]
LGTM, nice work!

File: dist/test_data.py [hunk 16]
```diff
@@ -260,4 +260,5 @@ def __ge__(self, other):
     'search_near_02' : Method(test_config),
     'search_near_03' : Method(test_config),
     'test_template' : Method(test_config),
+    'reverse_split' : Method(test_config),
```
Comment (etienneptl) [inline dist/test_data.py:263]: Looks like a missing A->Z 😉

Comment (luke-pearson) [reply dist/test_data.py]: [ef0a708](https://github.com/wiredtiger/wiredtiger/pull/8554/commits/ef0a7084082cda3917b66d9453de9276b0151389) I also added a comment explaining what that section is and that it should be alphabetically ordered.

**Review by etienneptl — APPROVED** [review]

**Review by jeremythorp — APPROVED** [review]
LGTM!

**Review by etienneptl — APPROVED** [review]

**Comment by luke-pearson** [comment]
evergreen merge

**Pull request merged**